### PR TITLE
AB Micro: publish assignment

### LINKED ISSUE #680: Microscopy: Publish experiment
as an instructor, I need to be able to publish my microscopy experiment so that students can access it. 

This issue includes any updates to the experiment model used by students.

### PR DESCRIPTION
**Pull request by annagav**
#### What are the relevant tickets?

Fix #680 
#### What's this PR do?

As an instructor, I can create an assignment with Microscopy technique and publish it.
As a student, I can login with course code and perform the experiment.

### CODE REVIEW

**Comment by pdpinch** [comment]
@annagav can this be reviewed before the filters are done (#697 / PR #705)?

Also, can you update the PR description?

**Comment by pdpinch** [comment]
Maybe this isn't quite ready, but when I attempt the published experiment as a student and I click the "New Microscopy" button, nothing happens. Looking in the JavaScript console, I see the following:

```
{"view":"microscopy","experiment_id":"scb_Experiment_11a31qgc28154","assignment_id":"a_8","onhashchange":true}
MainFrame.js?_=1471829212:629 {"view":"microscopy","experiment_id":"scb_Experiment_11a31qgc28154","assignment_id":"a_8","microscopy_id":"scb_Microscopy_ldkdm4286km3k","onhashchange":true}
raven.min.js:2 Uncaught TypeError: Cannot read property 'SP_ID_36' of undefined
```

**Comment by annagav** [comment]
I think the soy files need to recompiled. Run html_app/build.py

File: instructor/compiler.py [hunk 1]
```diff
@@ -303,39 +303,34 @@ def facs_kinds(assignment):
...
+                        'slide_type': sample_prep.micro_analysis
+                    }
+
+                for image in mapping.images.all():
+                    ret[key]['slides'].append(
                         {
-                            'hash': "IMAGE_{}".format(i.pk),
-                            'if_type': i.filter,
-                            'mag': i.objective
+                            'hash': image.pk,
```
Comment (noisecapella) [inline instructor/compiler.py:322]: What's happening in this change?

File: instructor/compiler.py [hunk 2]
```diff
@@ -406,13 +401,13 @@ def gel_types(a):
...
-        analysis = sp.analysis
-        condition = sp.condition
+    ret = {'na': {'name': 'None'}}
+    for sample_prep in a.microscopy_sample_prep.all():
+        analysis = sample_prep.micro_analysis
+        condition = sample_prep.condition
         if analysis not in ret:
             ret[analysis] = {
-                'name': analysis,
+                'name': sample_prep.get_micro_analysis_display(),
```
Comment (noisecapella) [inline instructor/compiler.py:410]: Where does this function come from?

File: instructor/compiler.py [hunk 3]
```diff
@@ -406,13 +401,13 @@ def gel_types(a):
 
 
 def micro_kinds(a):
-    ret = {}
-    for sp in a.microscopy_sample_prep.all():
-        analysis = sp.analysis
-        condition = sp.condition
+    ret = {'na': {'name': 'None'}}
```
Comment (noisecapella) [inline instructor/compiler.py:404]: What does `na` mean here?

**Comment by noisecapella** [comment]
Just some questions, functionality looks good

File: instructor/compiler.py (hunk 3 above)
Comment (annagav) [reply instructor/compiler.py:404]: not available

File: instructor/compiler.py (hunk 1 above)
Comment (annagav) [reply instructor/compiler.py:322]: the hash should be just the pk of the image. `if_type` is for the filters, that are not implemented yet.

File: instructor/compiler.py (hunk 2 above)
Comment (annagav) [reply instructor/compiler.py:410]: it is built in https://docs.djangoproject.com/en/1.10/ref/models/instances/#extra-instance-methods

**Comment by annagav** [comment]
@noisecapella, addressed all comments!

File: instructor/compiler.py (hunk 2 above)
Comment (noisecapella) [reply instructor/compiler.py:410]: Oh interesting

**Comment by noisecapella** [comment]
:+1:

**Pull request merged**